src/function/decoding_function.cpp

```diff
@@ -0,0 +1,145 @@
+				remaining_bytes_size = 1;
+				return;
+			}
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0x80 | (ch & 0x3F));
+		} else {
+			// 3-byte UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0xE0 | (ch >> 12));
+			if (decoded_buffer_current_position == decoded_buffer_size) {
+				// We are done, but we have to store two bytes for the next chunk!
+				encoded_buffer_current_position += 2;
+				remaining_bytes_buffer[0] = static_cast<char>(0x80 | ((ch >> 6) & 0x3F));
+				remaining_bytes_buffer[1] = static_cast<char>(0x80 | (ch & 0x3F));
+				remaining_bytes_size = 2;
+				return;
+			}
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0x80 | ((ch >> 6) & 0x3F));
+			if (decoded_buffer_current_position == decoded_buffer_size) {
+				// We are done, but we have to store one byte for the next chunk!
+				encoded_buffer_current_position += 2;
+				remaining_bytes_buffer[0] = static_cast<char>(0x80 | (ch & 0x3F));
+				remaining_bytes_size = 1;
+				return;
+			}
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0x80 | (ch & 0x3F));
+		}
+	}
+}
+
+void DecodeLatin1(char *encoded_buffer, idx_t &encoded_buffer_current_position, const idx_t encoded_buffer_size,
+                  char *decoded_buffer, idx_t &decoded_buffer_current_position, const idx_t decoded_buffer_size,
+                  char *remaining_bytes_buffer, idx_t &remaining_bytes_size) {
+	for (; encoded_buffer_current_position < encoded_buffer_size; encoded_buffer_current_position++) {
+		if (decoded_buffer_current_position == decoded_buffer_size) {
+			// We are done
+			return;
+		}
+		const unsigned char ch = static_cast<unsigned char>(encoded_buffer[encoded_buffer_current_position]);
+		if (ch > 0x7F && ch <= 0x9F) {
+			throw InvalidInputException("File is not latin-1 encoded");
+		}
+		if (ch <= 0x7F) {
+			// ASCII: 1 byte in UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(ch);
+		} else {
+			// Non-ASCII: 2 bytes in UTF-8
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>(0xc2 + (ch > 0xbf));
+			if (decoded_buffer_current_position == decoded_buffer_size) {
+				// We are done, but we have to store one byte for the next chunk!
+				encoded_buffer_current_position++;
+				remaining_bytes_buffer[0] = static_cast<char>((ch & 0x3f) + 0x80);
+				remaining_bytes_size = 1;
+				return;
+			}
+			decoded_buffer[decoded_buffer_current_position++] = static_cast<char>((ch & 0x3f) + 0x80);
+		}
+	}
+}
+
+void DecodeUTF8(char *encoded_buffer, idx_t &encoded_buffer_current_position, const idx_t encoded_buffer_size,
+                char *decoded_buffer, idx_t &decoded_buffer_current_position, const idx_t decoded_buffer_size,
+                char *remaining_bytes_buffer, idx_t &remaining_bytes_size) {
+	// This function just copies.
+	for (; encoded_buffer_current_position < encoded_buffer_size; encoded_buffer_current_position++) {
```

Comment:
Should we ever be calling this function to begin with? Perhaps `utf8` should not have a decode function, and we should check this one level up to ensure we're not doing an unnecessary copy?